src/keys/drivers/YubiKey.cpp

```diff
@@ -54,161 +132,162 @@ YubiKey* YubiKey::instance()
 {
-    m_mutex.lock();
-
-    // previously initialized
-    if (m_yk != nullptr && m_ykds != nullptr) {
-
-        if (yk_get_status(m_yk, m_ykds)) {
-            // Still connected
-            m_mutex.unlock();
-            return true;
-        } else {
-            // Initialized but not connected anymore, re-init
-            deinit();
-        }
-    }
+    return m_initialized;
+}
 
-    if (!yk_init()) {
-        m_mutex.unlock();
-        return false;
+void YubiKey::findValidKeys()
+{
+    m_error.clear();
+    if (!isInitialized()) {
+        return;
     }
 
-    // TODO: handle multiple attached hardware devices
-    m_onlyKey = false;
-    m_yk_void = static_cast<void*>(yk_open_first_key());
-#if YKPERS_VERSION_NUMBER >= 0x011400
-    // New fuction available in yubikey-personalization version >= 1.20.0 that allows
-    // selecting device VID/PID (yk_open_key_vid_pid)
-    if (m_yk == nullptr) {
-        static const int device_pids[] = {0x60fc}; // OnlyKey PID
-        m_yk_void = static_cast<void*>(yk_open_key_vid_pid(0x1d50, device_pids, 1, 0));
-        m_onlyKey = true;
-    }
-#endif
-    if (m_yk == nullptr) {
-        yk_release();
-        m_mutex.unlock();
-        return false;
-    }
+    QtConcurrent::run([this] {
+        if (!m_mutex.tryLock(1000)) {
+            emit detectComplete(false);
+            return;
+        }
 
-    m_ykds_void = static_cast<void*>(ykds_alloc());
-    if (m_ykds == nullptr) {
-        yk_close_key(m_yk);
-        m_yk_void = nullptr;
-        yk_release();
-        m_mutex.unlock();
-        return false;
-    }
+        // Remove all known keys
+        m_foundKeys.clear();
+
+        // Try to detect up to 4 connected hardware keys
+        for (int i = 0, j = 0; i + j < 4;) {
+            bool onlyKey = false;
+            auto* yk_key = openKey(i, j, &onlyKey);
+            Tools::wait(50);
+            if (yk_key) {
+                onlyKey ? ++j : ++i;
+                auto vender = onlyKey ? QStringLiteral("OnlyKey") : QStringLiteral("YubiKey");
+                auto serial = getSerial(yk_key);
+                if (serial == 0) {
+                    closeKey(yk_key);
+                    continue;
+                }
+
+                auto st = ykds_alloc();
+                yk_get_status(yk_key, st);
+                int vid, pid;
+                yk_get_key_vid_pid(yk_key, &vid, &pid);
+
+                auto chall = randomGen()->randomArray(1);
+                QByteArray resp;
+                QList<QPair<int, QString>> ykSlots;
+                for (int slot = 1; slot <= 2; ++slot) {
+                    auto config = (i == 1 ? CONFIG1_VALID : CONFIG2_VALID);
+                    if (ykds_touch_level(st) & config) {
+                        // Don't actually challenge a YubiKey Neo or below, they always require button press
+                        // if it is enabled for the slot resulting in failed detection
+                        if (pid <= 0x0116) {
```

Comment:
1) Magic number should be named constexpr (can be local).

2) Skipping the challenge results in both slots being detected as being configured for HMAC-SHA1 on my NEO, even though only one is. The other one is configured for OpenPGP. If I select that slot, KeePassXC will try to challenge it and then bails out with a timeout error. That's not how this should work and looks quite dangerous to me (particularly during master key setup).

Reply:
It's not dangerous because you cannot possibly get past the master key setup unless challenge response is setup correctly. The checks done in this function do not work with the NEO because of a firmware problem. The NEO does not cancel the CR action properly if "press" is enabled resulting in it waiting for a press when none will come. To add insult to injury, the NEO firmware also does not populate the touch status properly so I can't even  check if touch is enabled on a slot. The NEO is a fucking mess and I recommend just throwing it away and get a YK 4 or 5.

Reply:
It has served me well so far and I don't really want to buy a new one for no apparent reason. Having the other slot enabled all of a sudden is super ugly and annoying.
And what's dangerous is not the setup wizard, it's the master key change. When I tested it, I could change the master key and the error occurred only after that. I did not test what would have happened had closed the database in this state, but it certainly looks to me like a recipe for disaster or at least something that is hard to test yet easy to fuck up massively.

Reply:
I'll look into that further.

Reply:
> The NEO is a fucking mess and I recommend just throwing it away and get a YK 4 or 5.

Now I have noticed this... Ordering a replacement.